### DISCUSSION TITLE
Destroying Duplicate IKE_SA with Different Label

### DISCUSSION THREAD

**Discussion by andrew-hoff** [strongswan/strongswan#1040] [Q&A]
I am attempting to initiate two SAs between two peers over separate links. I have set `<conn>.unique` to `no`. Each peer uses the same Ike ID for each connection. As a result, I need to use `label` and `label_mode` to avoid a collision between the two connections. I have set the labels on both sides and confirmed that they were applied using `swanctl --list-conns`. I have set `label_mode` to `simple`. I am running StrongSwan version 5.9.6. However, I am still getting a `destroying duplicate IKE_SA for peer ..., received INITIAL_CONTACT` error.

Is there anything else I need to configure to get labels working?

**Comment by tobiasbrunner** [comment]
> As a result, I need to use label and label_mode to avoid a collision between the two connections.

Can you can distinguish the connections by IP (address/range/subnet)? And if you control both ends, what's the reason you can't use distinct identities?

The labels won't actually help you if you configure separate IKE_SAs. They are only relevant to select child configs of a particular connection, they are not used to select connections (that's what the IPs and identities are used for).

> However, I am still getting...

Because that has nothing to do with labels. Did you read the documentation for the `unique` setting?

**Comment by andrew-hoff** [comment]
Ah, that's helpful. I can try to re-architect to use a shared IKE_SA with different children instead of my current approach of using multiple IKE_SAs. I'd still like to use labels, however, because NAT makes it difficult to distinguish based on IP address/range/subnet.

I had read the documentation regarding the `unique` setting, though I must have misread it because I thought the labels also distinguished IKE_SAs.

Also, to answer your question, the reason why I can't use distinct identities is because I want to use the same certificate for both connections and my reading of the documentation implies that I can't use distinct identities for the same certificate.

**Reply by tobiasbrunner** [reply]
> I can try to re-architect to use a shared IKE_SA with different children instead of my current approach of using multiple IKE_SAs.

I may have misunderstood you, but didn't you want to have redundant IKE and CHILD_SAs over different links?

> I had read the documentation regarding the `unique` setting

Didn't you see then that you could have avoided that duplicate check by setting it to `never` so `INITIAL_CONTACT` notifies are ignored.

> I must have misread it because I thought the labels also distinguished IKE_SAs.

Which one? That of `unique` (which doesn't mention labels), or that of `label_mode` (which doesn't mention IKE_SAs or `unique`)? Let me know if you have suggestions to improve the docs.

> I want to use the same certificate for both connections and my reading of the documentation implies that I can't use distinct identities for the same certificate.

Which part of the documentation? Because you can easily do that, just add multiple identities to the certificates as subjectAltNames (SAN, e.g. via `--san` when using [`pki`](https://docs.strongswan.org/docs/5.9/pki/pki.html)). Then you can use a different identity for each connection.

**Reply by andrew-hoff** [reply]
> I may have misunderstood you, but didn't you want to have redundant IKE and CHILD_SAs over different links?

That's downstream of the real requirement. I suspect there are multiple configurations that would allow me to achieve my desired objective. Basically, I want to be able to use a route based VPN using GRE tunnels. However, I may be able to route to the same peer over different network interfaces. As a result, I create a GRE tunnel per peer + network interface combo. In other words, if I can route to peer 1 over network interfaces A and B, I will create two GRE tunnels with two GRE keys so that I can switch between paths by simply changing the route in my routing table to use whichever GRE tunnel I desire.

To apply the appropriate encryption, I specify the GRE key in the traffic selectors of the Child SA. This way, when I route through a given GRE tunnel, the appropriate key is applied, which matches the xfrm policy, so the appropriate encryption is applied.

I do not think that the requirement above necessitates that I use two IKE SAs for each peer + network interface combo. A few ideas come to mind that work around this problem:
- I can use a single IKE SA per peer with a Child SA per network interface through which I can reach the peer. Each child's traffic selector will have the appropriate GRE key so that packets sent through the corresponding GRE tunnel have the appropriate encryption applied.
- I can use a single IKE SA and a single Child SA per peer, but extend my traffic selectors to support multiple GRE keys.
- I can share a GRE key across all GRE tunnels for the same peer.

For each of these, I'll need to better understand expected behavior when a given route goes down. For example, if I can reach peer 2 from peer 1 over network interfaces A and B and initiate the connection over A, what happens if network interface A goes down 20 minutes later? I'll need to do some digging through the documentation and code to better answer this question.

> Didn't you see then that you could have avoided that duplicate check by setting it to never so INITIAL_CONTACT notifies are ignored.

If I do this, what happens if my peer is restarted and that peer is the initiator? In that case, will I not have two IKE SAs for that same peer? That was my interpretation of the documentation.

> Which one? That of unique (which doesn't mention labels), or that of label_mode (which doesn't mention IKE_SAs or unique)? Let me know if you have suggestions to improve the docs.

I think I misinterpreted the `label_mode` config when I read 'the label will be used as is as an additional identifier/selector'. After reading your response, I think the misinterpretation was my fault for not properly differentiating between a connection and a Child SA.

> Which part of the documentation? Because you can easily do that, just add multiple identities to the certificates as subjectAltNames (SAN, e.g. via --san when using pki). Then you can use a different identity for each connection.

Based on the setup I described above, I think this would be tricky to implement. We don't know up front how many network interfaces a node is going to have, so we won't know how many subjectAltNames to create in the certificate. Basically, I want to avoid coupling the contents of the certificate with the connections we may or may not form.
